Hw1.1

Goodwin999 opened the pull request.

Goodwin999 committed:
first step completed

Goodwin999 committed:
определил тело метода create для Student;

Goodwin999 committed:
создал класс для исключений,возникающих в результате выполнения метода delete, определил тело метода

Goodwin999 committed:
определил логику методов класса FacultyServiceImpl;
создал контроллеры и эндпоинты для сущностей типа Faculty и Student

Goodwin999 committed:
выполнил шаг 5:
Добавить фильтрацию студентов по возрасту.
Добавить фильтрацию факультетов по цвету.

Goodwin999 committed:
IDEA запускает приложение, на Swegger не работает, не могу разобраться.

Goodwin999 committed:
Рабочий Вариант

Goodwin999 committed:
Добавил БД.

Goodwin999 committed:
В файл application.properties добавлены четыре свойства. Приложение успешно запускается без ошибок

Goodwin999 committed:
Шаг 3 выполнен

Goodwin999 committed:
Шаг 4 выполнен

Goodwin999 committed:
Шаг 5 выполнен. Выполнил ДЗ

dmitry-bizin commented on pom.xml:
```diff
@@ -16,7 +16,31 @@
 	</properties>
+	<repositories>
+		<repository>
+			<id>jakarta.ee</id>
```
это не нужно, у нас нет Jakarta EE

dmitry-bizin commented on pom.xml:
```diff
@@ -16,7 +16,31 @@
 	<dependencies>
+		<dependency>
+			<groupId>jakarta.servlet</groupId>
+			<artifactId>jakarta.servlet-api</artifactId>
```
зависимость не нужна, у нас нет Jakarta EE

dmitry-bizin commented on pom.xml:
```diff
@@ -16,7 +16,31 @@
+
+		<dependency>
+			<groupId>io.swagger.core.v3</groupId>
+			<artifactId>swagger-annotations</artifactId>
```
это тоже не нужно, openapi подключена ниже

dmitry-bizin commented on pom.xml:
```diff
@@ -27,6 +51,16 @@
+		<dependency>
+			<groupId>org.postgresql</groupId>
+			<artifactId>postgresql</artifactId>
+			<version>42.6.0</version>
```
версию писать не нужно

dmitry-bizin commented on src/main/java/ru/hogwarts/school/SchoolApplication.java:
```diff
@@ -1,9 +1,14 @@
+
 @SpringBootApplication
+@OpenAPIDefinition
+@SecurityRequirement(name = "bearerAuth")
```
это не нужно, аутентификации в приложении нет никакой

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,76 @@
+public class FacultyController {
+    private final FacultyService facultyService;
+
+    @Autowired
```
аннотация не нужна

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Operation(summary = "Создание нового факультета")
+    @PostMapping("/create")
```
никаких слов в пути URL быть не должно, это нарушает REST, т.к. сам метод POST уже говорит о том, что это создание, поэтому если ещё писать слова в пути, то это будет тавтология

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Operation(summary = "Получение факультета по ID")
+    @GetMapping("/read/{id}")
```
никаких слов в пути URL быть не должно, это нарушает REST, т.к. сам метод GET уже говорит о том, что это чтение, поэтому если ещё писать слова в пути, то это будет тавтология

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Operation(summary = "Обновление информации о факультете")
+    @PutMapping("/update")
```
никаких слов в пути URL быть не должно, это нарушает REST, т.к. сам метод PUT уже говорит о том, что это обновление, поэтому если ещё писать слова в пути, то это будет тавтология

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Operation(summary = "Удаление факультета по ID")
+    @DeleteMapping("/delete/{id}")
```
никаких слов в пути URL быть не должно, это нарушает REST, т.к. сам метод DELETE уже говорит о том, что это удаление, поэтому если ещё писать слова в пути, то это будет тавтология

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Operation(summary = "Фильтрация факультетов по цвету")
+    @GetMapping("/filterByColor/{color}")
```
никаких слов в пути URL быть не должно, это нарушает REST. Здесь просто должен быть корень, а параметр должен быть параметром запроса, а не пути. В пути можно передавать только идентификатор, иначе  это снова будет нарушать принципы REST

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,76 @@
+        List<Faculty> filteredFaculties = facultyService.filterFacultiesByColor(color);
+        if (filteredFaculties.isEmpty()) {
+            String message = "Факультеты с цветом " + color + " не найдены";
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(message);
```
это неверно, надо просто вернуть пустую коллекцию с кодом 200, а не 404. Код 404 должен быть возвращён тогда и только тогда, когда идёт обращение к ресурсу, который не найден. Здесь обращения к ресурсу нет

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -0,0 +1,69 @@
+@Tag(name = "Управление студентами", description = "Методы для работы со студентами")
+public class StudentController {
+    private final StudentService studentService;
+
```
здесь замечания аналогичные как для контроллера факультетов

dmitry-bizin commented on src/main/java/ru/hogwarts/school/exception/DatabaseAccessException.java:
```diff
@@ -0,0 +1,34 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
```
```suggestion
@RestControllerAdvice
```

dmitry-bizin commented on src/main/java/ru/hogwarts/school/exception/DatabaseAccessException.java:
```diff
@@ -0,0 +1,34 @@
+
+@ControllerAdvice
+
+public class DatabaseAccessException extends RuntimeException {
```
Если этот класс является исключением, то никакой аннотации и метода внутри него быть не должно. Если же этот класс является обработчиком исключений, то никакого наследования и конструкторов быть не должно. Ты просто взял и собрал в кучу исключение и обработчик, что неверно

dmitry-bizin commented on src/main/java/ru/hogwarts/school/model/Faculty.java:
```diff
@@ -0,0 +1,81 @@
+    private Long id;
+    private String name;
+    private String color;
+    @OneToMany(fetch = FetchType.LAZY, mappedBy = "faculty")
+    private List<Student> studentList;
```
это поле здесь не нужно, иначе будет бесконечная рекурсия при получении факультетов, связанных со студентами

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyServiceImpl.java:
```diff
@@ -0,0 +1,70 @@
+@Service
+public class FacultyServiceImpl implements FacultyService {
+    private final FacultyRepository facultyRepository;
+
+@Autowired
```
аннотация не нужна

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyServiceImpl.java:
```diff
@@ -0,0 +1,70 @@
+
+    @Override
+    public Faculty create(Faculty faculty) {
+        Optional<Faculty> existingFaculty = facultyRepository.findAll().stream()
+                .filter(f -> f.getName().equalsIgnoreCase(faculty.getName()))
+                .findFirst();
+
+        if (existingFaculty.isPresent()) {
+            throw new IllegalArgumentException("Факультет с таким именем уже существует.");
+        }
```
это неверно, никакой проверки при создании быть не должно, создание - это просто сохранение, но надо обnullить id

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyServiceImpl.java:
```diff
@@ -0,0 +1,70 @@
+    @Override
+    public Faculty update(Faculty faculty) {
+        if (!facultyRepository.existsById(faculty.getId())) {
+            throw new IllegalArgumentException("Факультет с id " + faculty.getId() + " не найден.");
```
исключение здесь кидать не нужно, у тебя выше в контроллере ожидается null, см. внимательнее

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyServiceImpl.java:
```diff
@@ -0,0 +1,70 @@
+    @Override
+    public boolean delete(long id) {
+        if (!facultyRepository.existsById(id)) {
+            throw new IllegalArgumentException("Факультет с id " + id + " не найден.");
```
исключение здесь кидать не нужно, у тебя выше в контроллере ожидается false, см. внимательнее

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyServiceImpl.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public List<Faculty> filterFacultiesByColor(String color) {
+        List<Faculty> allFaculties = facultyRepository.findAll();
+        List<Faculty> filteredFaculties = new ArrayList<>();
+
+        for (Faculty faculty : allFaculties) {
+            if (faculty.getColor().equalsIgnoreCase(color)) {
+                filteredFaculties.add(faculty);
+            }
+        }
```
Это неверно. Вот представь, что у тебя в таблице тысяча строк, ты все эти строки получаешь в виде объектов, а затем оставляешь только малую часть, отфильтровывая. У тебя будет либо OutOfMemoryError или просто приложение будет тормозить. Никогда не надо использовать findAll, это почти всегда ошибка, все фильтрации надо делать на уровне запроса к БД

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/StudentServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+public class StudentServiceImpl implements StudentService {
+
+    private final StudentRepository studentRepository;
+
```
здесь замечания как для сервиса факультетов

Goodwin999 committed:
Исправил все недостатки и замечанияя

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -40,37 +33,28 @@ public ResponseEntity<Faculty> getFaculty(@PathVariable long id) {
-    @Operation(summary = "Фильтрация факультетов по цвету")
-    @GetMapping("/filterByColor/{color}")
-    public ResponseEntity<?> filterFacultiesByColor(@PathVariable String color) {
+    @GetMapping("/filter")
```
```suggestion
    @GetMapping(params = "color")
```
не должно быть никаких слов в пути

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -17,53 +15,43 @@
-    @GetMapping("/filterByAge/{age}")
-    public ResponseEntity<List<Student>> filterStudentsByAge(@PathVariable int age) {
+
+    @GetMapping("/filter")
```
```suggestion
    @GetMapping(params = "age")
```
не должно быть никаких слов в пути

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyServiceImpl.java:
```diff
@@ -1,34 +1,30 @@
+        try {
+            faculty.setId(null);
+            return facultyRepository.save(faculty);
+        } catch (Exception ex) {
```
не нужно здесь ловить исключение, оно никогда не будет выброшено

Goodwin999 committed:
3.4. Введение в SQL
шаг 1 выполнил

Goodwin999 committed:
после выполнения 1 шага приложение падало сразу после запуска, проблему устранил.

Goodwin999 committed:
Выполнил 2 шаг: "Подключиться с помощью IDEA к базе данных и выполнить простой запрос select * from student".;
Шаг выполнен с помощью инструмента управления базой данных- "DBeaver";

Goodwin999 committed:
Выполнил 3 шаг.

Goodwin999 committed:
Выполнил 4 шаг. ДЗ выполнено, приложение каким-то образом запускается и не падает, был конфликт между ЮРЛ адресами, получилось устранить.

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -14,6 +16,7 @@
 @Tag(name = "Управление студентами", description = "Методы для работы со студентами")
 public class StudentController {
     private final StudentService studentService;
+    private StudentRepository studentRepository;
```
репозитории не должны использоваться в контроллерах, их надо использовать только в сервисах

Goodwin999 committed:
3.5. Потоки данных. Работа с файлами
Выполнен 1 шаг.
Приложение не падает.

Goodwin999 committed:
3.5. Потоки данных. Работа с файлами
Выполнен 2 шаг.
Приложение не падает.

Goodwin999 committed:
3.5. Потоки данных. Работа с файлами
Выполнен 3 шаг.
Приложение все еще чудным образом не падает.

Goodwin999 committed:
3.5. Потоки данных. Работа с файлами
Выполнен 3 шаг.
Приложение все еще чудным образом не падает.

Goodwin999 committed:
3.5. Потоки данных. Работа с файлами.
Выполнены все шаги, провел небольшой рефакторинг, прошел все проверки через Swagger, фотографии сохраняются на диск, и в БД.

Goodwin999 committed:
Интеграционные тесты.Начало работы.Программа работает.Только создал папку test/java пометил как тест ресурс приложения.

Goodwin999 committed:
Ветка с с 1 заданием по тестам;
создал первый тест метода "создать студента", добавил аннотацию контекст конфигурейшн, так как тест падал при первом запуске из за невозможности найти основной класс;

Goodwin999 committed:
Задание 1 выполнил не полностью, не смог найти причину, почему падают некоторые тесты, по ошибкам показывает что ответы возвращаются в неправильном формате, причину так и не понял.

Goodwin999 committed:
Задание 2;Начало;

Goodwin999 committed:
Засада;Боль;Тлен;Ненависть;ничего не получислось

Goodwin999 committed:
Все работает в тестах контроллера студентов, кроме последнего теста, пока что не нашел причину.

Goodwin999 committed:
Тоже самое, в обоих классах тестов, не работает метод, связанный с поиском студента по факультету и факультета по студенту, пока что не разобрался почему так

Goodwin999 committed:
FacultyControllerMvcTest поднялся все тесты заработали, добавил аннотации @JsonBackReference и  @JsonManagedReference для сериализации и решения проблемы бесконечной рекурсии.
слабо понимаю что происходит

Goodwin999 committed:
FacultyControllerMvcTest поднялся все тесты заработали, добавил аннотации @JsonBackReference и  @JsonManagedReference для сериализации и решения проблемы бесконечной рекурсии.
слабо понимаю что происходит

Goodwin999 committed:
Поменял адрес теста testGetStudentFaculty в тестах, для запроса факультета, теперь все тесты MVC поднимаются

Goodwin999 committed:
РАбочий вариант MvC тестов

Goodwin999 committed:
Поправил URL контроллеров

Goodwin999 committed:
Поправил URL контроллеров(убрал слэш из студентКонтроллер)

Goodwin999 committed:
Поправил URL контроллеров(убрал слэш из студентКонтроллер)

Goodwin999 committed:
Поправил URL контроллеров, параметры энпоинта getStudentFaculty, поставил Id, тест работает, я молодец

Goodwin999 committed:
Попровил энпоинт получение студентов по факультету, сваггер работает, студенты факультета возвращаются;

добавили метод для получения студентов факультета в реализацию сервиса;

Добавили аннотации, для предотвращения рекурсии между студентами и факультетами;

Goodwin999 committed:
Поправил поля тестовых классов добавил репозитории, и сервисы;

Goodwin999 committed:
FacultyControllerTestRestTemplate, все тесты поднимаются;

Goodwin999 committed:
FacultyControllerTestRestTemplate, все тесты поднимаются;

Goodwin999 committed:
Merge remote-tracking branch 'origin/WebMVcTestBranchApp' into HW1.1

# Conflicts:
#	src/main/java/ru/hogwarts/school/model/Student.java

Goodwin999 committed:
Сделал слияние веток, по прежнему не работает один тест на получение факультета по айди студента, приложение посыпалось

Goodwin999 committed:
4.1 Начало работы, добавление нового функционала;
Создал методы в репозитории студентов;

Goodwin999 committed:
Продолжение борьбы с тестами;

Goodwin999 committed:
Продолжение борьбы с тестами;
УСТАЛ

Goodwin999 committed:
Продолжение борьбы с тестами;
Откатил изменения контроллера

Goodwin999 committed:
ДЗ по Пагинации Аватарок и новых контроллеров сделал!

Goodwin999 committed:
Поправил аватар контроллер, сервис и его реализацию (имплементацию)

Goodwin999 committed:
Поправил аватар контроллер, сервис и его реализацию (имплементацию);
Дополнительно поправил метод контроллера добавил Pageable (Не знаю зачем)

Goodwin999 committed:
Управление схемами баз данных, начало работы;
Команды все функционируют в DBeaver, файл 421.sql команды подчеркивает красным, как выяснил, моя версия IDE не поддерживат такие вещи;

Goodwin999 committed:
Управление схемами баз данных;
Шаг 2 выполнен, таблицы с зависимостями созданы, команды выполнены через консоль Dbeaver

Goodwin999 committed:
Управление схемами баз данных;
Шаг 3 выполнен, ДЗ сделал;

Goodwin999 committed:
Миграции и индексы, начало работы;
Добавлена зависимость в pom.xml и путь в app.properties;

Goodwin999 committed:
Миграции и индексы, начало работы;
Изменил название файа миграции согласно задания;

Goodwin999 committed:
Миграции и индексы, Шаг 2 выполнен, ДЗ выполнено

Goodwin999 committed:
Миграции и индексы, Шаг 2 выполнен, ДЗ выполнено;
Поправил код запроса, добавил проверки на наличие выполненных индексов т.к. приложение не запускалось из-за того, что я предварительно выполнял команды в Dbeaver

Goodwin999 committed:
Логирование и конфигурирование в приложении;
Начало работы, добавил зависимость в помник для логов;

Goodwin999 committed:
Логирование и конфигурирование в приложении;
Начало работы, Настроил уровни логгирования ддля приложения, добавил вывод сообщения в логи для метода create у студентов;

Goodwin999 committed:
Логирование и конфигурирование в приложении;
Шаг 1 - выполнен;

Goodwin999 committed:
Логирование и конфигурирование в приложении;
Шаг 2 - выполнен;

Goodwin999 committed:
Логирование и конфигурирование в приложении;
Шаг 3 - выполнен;

Goodwin999 committed:
Логирование и конфигурирование в приложении;
Шаг 3 - выполнен;

Goodwin999 committed:
Логирование и конфигурирование в приложении;
Шаг 3 - выполнен;

AleksandrZhukovJava commented on pom.xml:
```diff
@@ -1,41 +1,59 @@
+        <dependency>
+            <groupId>org.springdoc</groupId>
+            <artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
+            <version>2.1.0</version>
```
для такого бута старовато чуть чуть, бери лучше 2.6.0 и выше

AleksandrZhukovJava commented on src/main/java/ru/hogwarts/school/controller/AvatarController.java:
```diff
@@ -0,0 +1,127 @@
+            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(null);
+        }
+        try {
+            Avatar avatar = new Avatar();
+            avatar.setFilePath(file.getOriginalFilename());
+            avatar.setFileSize(file.getSize());
+            avatar.setMediaType(file.getContentType());
+            avatar.setData(file.getBytes());
+            Student student = studentService.read(studentId);
+            if (student == null) {
+                return ResponseEntity.badRequest().body(null);
+            }
+            avatar.setStudent(student);
+            avatar = avatarService.saveAvatar(avatar);
+            String fileName = StringUtils.cleanPath(file.getOriginalFilename());
+            System.out.println("Имя файла: " + fileName);
+            Path filePath = Paths.get(uploadDir, fileName);
+            System.out.println("Путь для сохранения файла: " + filePath.toString());
+
+            try (FileOutputStream fos = new FileOutputStream(filePath.toFile())) {
+                System.out.println("Начало записи файла...");
+                fos.write(file.getBytes());
+                System.out.println("Файл успешно сохранен: " + filePath.toString());
+            } catch (IOException e) {
+                System.err.println("Ошибка при сохранении файла: " + e.getMessage());
+                return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(null);
+            }
+            AvatarDTO avatarDTO = new AvatarDTO();
+            avatarDTO.setId(avatar.getId());
+            avatarDTO.setFilePath(avatar.getFilePath());
+            avatarDTO.setFileSize(avatar.getFileSize());
+            avatarDTO.setMediaType(avatar.getMediaType());
```
это всё должно быть в сервисе, бизнес логики в контроллере быть не должно

AleksandrZhukovJava commented on src/main/resources/application.properties:
```diff
@@ -1 +1,12 @@
+spring.liquibase.enabled=true
+spring.liquibase.change-log=classpath:liquibase/changelog-master.yml
+
+spring.profiles.active=prod
```
профайл явно не задают, его нужно задавать при запуске приложения в конфигурациях

Goodwin999 committed:
4.5 Параллельные стримы
Начало работы, шаг 1 выполнен

Goodwin999 committed:
4.5 Параллельные стримы
Шаг 2 выполнен, реализован функционал для получаения среднего возраста студентов через Stream API, при этом метод репозитория, реализующий предопределенный SQL запрос оставлен для стабильности функционала, добавлен контроллер и реализация метода для работы со стримами.
Согласно прошлым заданиям проекта, метод для получения ср возраста студентов уже существовал;

Goodwin999 committed:
4.5 Параллельные стримы
Шаг 3 выполнен

Goodwin999 committed:
4.5 Параллельные стримы
Шаг 4 выполнен;
Сделал класс-сервис с реализацией, без интерфейса, создан контроллер, возвращающий результат согласно задания;
Приложение работает;

ДЗ выполнено

Goodwin999 committed:
4.5 Параллельные стримы
Доработал ошибки

Goodwin999 committed:
4.6 потоки
ДЗ выполнено, все шаги пройдены

Goodwin999 committed:
4.6 потоки
ДЗ выполнено,
все шаги пройдены, приложение работает;

Goodwin999 committed:
4.6 потоки
ДЗ выполнено,
все шаги пройдены, приложение работает;

Goodwin999 committed:
4.6 потоки
Не много поправил проект, поменял конфиг файлы как положено;

AleksandrZhukovJava commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -108,16 +108,47 @@ public ResponseEntity<List<Student>> getLastFiveStudents() {
+        thread2.start();
+        return ResponseEntity.ok("Student names are printed in synchronized mode.");
 
+    }
```
А почему в контроллере то? Всю логику, даже тестовую, мы пишем в сервисах
